### DISCUSSION TITLE
QUploader: extract filename from 'uploaded' event

### DISCUSSION THREAD

**Discussion by hhaensel** [quasarframework/quasar#17187] [General - Components / Directives / etc] [answered]
I'm trying to use the 'uploaded' of QUploader event to trigger some action on the file.
Currently, the event only transmits the file.__key property plus some stats but not the name. The __key, however does not allow to extract the filename if the file starts with a number.

https://github.com/quasarframework/quasar/blob/4207741e1df61079c5da6bb944dc754a34c203cf/ui/src/composables/private.use-file/use-file.js#L130

So I currently include a ref to the uploader component and use `findIndex` to determine correct file name.
Wouldn't it be a good idea to either add the filename in the emitted events or to use a separator like '-' or '_'

```js
file.__key = file.webkitRelativePath + file.lastModified + '-' + file.name + file.size
```

**Comment by yusufkandemir** [comment, marked as answer]
`file.__key` is a Quasar addition to the native `File` object, what you are looking for is [`File.name`](https://developer.mozilla.org/en-US/docs/Web/API/File/name).

**Reply by hhaensel** [reply]
Thanks for answering so quickly. It turned out that I forwarded the event object to my backend server but somehow the file.name was not included. Now that I used a client-side `console.log(event)`, I know that I must debug my event forwarding.
Thanks a lot!

**Reply by hhaensel** [reply]
I did my homework and found that the standard File object is not stringified by JSON. Adding
```js
File.prototype.toJSON = function() {
    return {
        lastModified: this.lastModified,
        name: this.name,
        size: this.size,
        type: this.type,
        webkitRelativePath: this.webkitRelativePath
    };
}
```
to my page solved the problem.

**Reply by hhaensel** [reply]
If you want to retain Quasar's user-defined props you can go with
```js
File.prototype.toJSON = function() {
    const imageObject = {};

    for (const key in this) {
        const value = this[key];
        (typeof value !== "function") && (imageObject[key] = value);
    }

    return imageObject;
}
```
